llvm/labs/lab2/akopyan_zal/akopyanInliningPass.cpp

```diff
@@ -0,0 +1,199 @@
+                             size_t &counter_inlined) {
+    for (BasicBlock &CalleeBB : *Callee) {
+      BasicBlock *NewBB = BasicBlock::Create(
+          F.getContext(),
+          CalleeBB.getName() + ".inlined." + std::to_string(counter_inlined),
+          &F);
+      BlockMap[&CalleeBB] = NewBB;
+    }
+    counter_inlined++;
+  }
+
+  void
+  updateTerminatorOfInsertBB(BasicBlock *InsertBB, Function *Callee,
+                             DenseMap<BasicBlock *, BasicBlock *> &BlockMap) {
+    InsertBB->getTerminator()->setSuccessor(0,
+                                            BlockMap[&Callee->getEntryBlock()]);
+  }
+
+  void
+  copyInstructionsFromCallee(Function &F, Function *Callee,
+                             DenseMap<BasicBlock *, BasicBlock *> &BlockMap,
+                             ValueToValueMapTy &VMap) {
+    for (BasicBlock &CalleeBB : *Callee) {
+      BasicBlock *NewBB = BlockMap[&CalleeBB];
+      for (Instruction &Inst : CalleeBB) {
+        IRBuilder<> Builder(NewBB);
+        Instruction *NewInst = Inst.clone();
+        Builder.Insert(NewInst);
+        VMap[&Inst] = NewInst;
+      }
+    }
+  }
+
+  void
+  updateBranchInstructions(Function &F,
+                           DenseMap<BasicBlock *, BasicBlock *> &BlockMap) {
+    for (BasicBlock &CalleeBB : F) {
+      for (Instruction &Branch : CalleeBB) {
+        if (auto *BI = dyn_cast<BranchInst>(&Branch)) {
+          for (size_t i = 0, e = BI->getNumSuccessors(); i != e; ++i) {
+            BasicBlock *Successor = BI->getSuccessor(i);
+            if (BlockMap[Successor] != nullptr) {
+              BI->setSuccessor(i, BlockMap[Successor]);
+            }
+          }
+        }
+      }
+    }
+  }
+
+  void remapInstructions(Function *Callee,
+                         DenseMap<BasicBlock *, BasicBlock *> &BlockMap,
+                         ValueToValueMapTy &VMap) {
+    for (BasicBlock &CalleeBB : *Callee) {
+      BasicBlock *NewBB = BlockMap[&CalleeBB];
+      for (Instruction &Inst : *NewBB) {
+        RemapInstruction(&Inst, VMap, RF_IgnoreMissingLocals, nullptr, nullptr);
+      }
+    }
+  }
+
+  void handleReturnInstructions(Function &F, BasicBlock *SplitBB) {
+    IRBuilder<> Builder(F.getContext());
+    for (BasicBlock &BB : F) {
+      SmallVector<ReturnInst *, 16> ReturnInstructions;
+      for (Instruction &I : BB) {
+        if (auto *RI = dyn_cast<ReturnInst>(&I)) {
+          ReturnInstructions.push_back(RI);
+        }
+      }
+
+      for (ReturnInst *RI : ReturnInstructions) {
+        if (RI->getParent() != SplitBB) {
+          RI->eraseFromParent();
+          Builder.SetInsertPoint(&BB);
+          Builder.CreateBr(SplitBB);
+        }
+      }
+    }
+  }
+
+  void removeCallInstruction(CallInst *CI) { CI->eraseFromParent(); }
+
+  void moveSplitBBAfterNextBlock(Function &F, BasicBlock *SplitBB) {
+    Function::iterator it = F.begin();
+    while (&*it != SplitBB && it != F.end()) {

```

Comment:
```suggestion
    while (it != F.end() && &*it != SplitBB) {
```
Dangerous operation. You should check first and only if it is available you can dereference it.